source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,194 @@
+  if (operand_index != 2 && operand_index != 3) {
+    return false;
+  }
+
+  // The new instruction value must exist and must be OpConstant.
+  auto new_instruction = ir_context->get_def_use_mgr()->GetDef(
+      message_.memory_semantics_new_value_id());
+  if (!new_instruction) {
+    return false;
+  }
+  if (new_instruction->opcode() != SpvOpConstant) {
+    return false;
+  }
+  auto new_memory_sematics_value = static_cast<SpvMemorySemanticsMask>(
+      new_instruction->GetSingleWordInOperand(0));
+
+  // Instruction must be atomic instruction only. Operand index must be equal to
+  // 2 in the case of instruction that takes one memory semantics operand.
+  // Instructions that takes two memory semantics, one of the value must be
+  // stronger than the other.
+  switch (needed_atomic_instruction->opcode()) {
+    case SpvOpAtomicLoad:
+    case SpvOpAtomicStore:
+    case SpvOpAtomicExchange:
+    case SpvOpAtomicIIncrement:
+    case SpvOpAtomicIDecrement:
+    case SpvOpAtomicIAdd:
+    case SpvOpAtomicISub:
+    case SpvOpAtomicSMin:
+    case SpvOpAtomicUMin:
+    case SpvOpAtomicSMax:
+    case SpvOpAtomicUMax:
+    case SpvOpAtomicAnd:
+    case SpvOpAtomicOr:
+    case SpvOpAtomicXor:
+    case SpvOpAtomicFlagTestAndSet:
+    case SpvOpAtomicFlagClear:
+    case SpvOpAtomicFAddEXT:
+      if (operand_index != 2) {
+        return false;
+      }
+      break;
+
+    case SpvOpAtomicCompareExchange:
+    case SpvOpAtomicCompareExchangeWeak:
+      if (new_memory_sematics_value ==
+          static_cast<SpvMemorySemanticsMask>(
+              ir_context->get_def_use_mgr()
+                  ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(
+                      operand_index == 3 ? 2 : 3))
+                  ->GetSingleWordInOperand(0))) {
+        return false;
+      }
+      break;
+
+    default:
+      return false;
+  }
+
+  // The new memory semantics value must be larger than old.
+  auto old_memory_sematics_value = static_cast<SpvMemorySemanticsMask>(
+      ir_context->get_def_use_mgr()
+          ->GetDef(
+              needed_atomic_instruction->GetSingleWordInOperand(operand_index))
+          ->GetSingleWordInOperand(0));
+  if (new_memory_sematics_value <= old_memory_sematics_value) {
+    return false;
+  }
+
+  if (!IsValidConverstion(needed_atomic_instruction->opcode(),
+                          new_memory_sematics_value)) {
+    return false;
+  }
+
+  return true;
+}
+
+void TransformationChangingMemorySemantics::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto needed_atomic_instruction =
+      FindInstruction(message_.atomic_instruction(), ir_context);
+  needed_atomic_instruction->SetInOperand(
+      message_.memory_semantics_operand_index(),
+      {message_.memory_semantics_new_value_id()});
+}
+
+bool TransformationChangingMemorySemantics::IsValidConverstion(
+    SpvOp opcode, SpvMemorySemanticsMask new_memory_sematics_value) {
```

Comment:
Something else I just noticed in the spec: If the "memory model" of the SPIR-V module is "Vulkan" then "SequentiallyConsistent" must not be used. So you will need to also check this, which means you will need to pass another parameter into this function (e.g. "is vulkan memory model" bool, the memory mode itself, or the `ir_context`).

Reply:
Done.